snuba/query/parser/__init__.py

```diff
@@ -19,9 +21,32 @@ def parse_query(body: MutableMapping[str, Any], dataset: Dataset) -> Query:
     Parses the query body generating the AST. This only takes into
     account the initial query body. Extensions are parsed by extension
     processors and are supposed to update the AST.
+
+    Parsing includes two phases. The first transforms the json body into
+    a minimal query Object resolving expressions, conditions, etc.
+    The second phase performs some query processing to provide a sane
+    query to the dataset specific section.
+    - It prevents alias shadowing.
+    - It transforms columns from the tags[asd] form into
+      SubscriptableReference.
+    - Applies aliases to all columns that do not have one and that do not
+      represent a reference to an alias.
+      During query processing a column can be transformed into a different
+      expression. It is essential to preserve the original column name so
+      that the result set still has a column with the name provided by the
+      user no matter on which transformation we applied.
+      By applying aliases at this stage every processor just needs to
+      preserve them to guarantee the correctness of the query.
     """
     try:
-        return _parse_query_impl(body, dataset)
+        query = _parse_query_impl(body, dataset)
+        # These are the post processing phases
+        _validate_aliases(query)
+        _parse_subscriptables(query)
+        _apply_column_aliases(query)
+        # WARNING: These steps above assume table resolution did not happen
```

Comment:
Is there anyway to enforce this via a test or something?

Reply:
I can add logging to ensure the sanity (and when we add table resolution we can enforce that it does not happen). Basically it checks that no table name is populated before we start here. (They won't because table resolution is not implemented at all yet)

Reply:
done